### ISSUE TITLE
Lazy loaded module in named outlet throws error

### ISSUE THREAD

**Issue by molcik** [angular/angular#12842]
[x] bug report 
[ ] feature request
[ ] support request 

**Current behavior**
Navigating to route with lazy loaded module in named outlet throws error: 

> Uncaught (in promise): TypeError: Cannot read property 'routes' of undefined

**Expected behavior**
Navigating to route with lazy loaded module in named outlet without errors.

**Minimal reproduction of the problem with instructions**
When I have configured my routes with named outlet and want to have lazy loaded module in that outlet. 

```TypeScript
export const routes: Routes = [
 {  
   path: '',
   component: AppComponent,
   children: [
     {
       path: '',
       loadChildren: "app/home/home.module#HomeModule"
     },   {
       path: '',
       loadChildren: "app/about/about.module#AboutModule",
       outlet: "about"
     }
   ]
  }
];
```
Here is the plnkr to demonstrate the behavior: http://plnkr.co/edit/nK9kJYdakXILZU0UKFrP?p=preview 
(Anything is displayed because of error)

**What is the motivation / use case for changing the behavior?**
Using lazy loaded modules in named outlets.

Please tell us about your environment:

Angular version: 2.0.X

Browser: [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Language: [all | TypeScript X.X | ES6/7 | ES5]

Node (for AoT issues): node --version =

**Comment by CalvinDale** [comment]
This issue also affects updating matrix-params via `router.navigate()`. E.g.,

```
  {
    path: '', component: DummyWrapperComponent, // Wrapped because a component-less route cannot have a named outlet
    outlet: 'popup', // Targetting `popup` outlet in wrapper's parent
    children: [
      {
        path: 'state-scratchpad',
        // outlet: 'popup', // Can't target non-primary outlet for lazy-loaded component
        // params: { tcomp: '', tmake: '' }, // These matrix params can be updated via router.navigate() if targeting
        //     `primary` outlet without the wrapper; but for `popup` outlet the wrapper is required
        loadChildren: 'app/state-scratchpad/state-scratchpad.module#StateScratchpadModule'
      }
    ]
  }
```

**Comment by pleerock** [comment]
have anyone find a temporary workaround for this issue? (goal is to load component in named outlets lazily).

**Comment by pleerock** [comment]
@UnwrittenFun yeah it works however it brings a side effect - extra level in the component tree which can bring some problems in some use cases

**Comment by RatkoR** [comment]
I know It's been a while since commit 8f2fa0f766d9f, but can you @vsavkin remember why you added that "a componentless route cannot have a named outlet set" guard. 

Is it because it's just not possible to lazy load in such a way (because maybe router is designed so) or is this something that will eventually be resolved sometime.

**Comment by KhanaevRoman** [comment]
@UnwrittenFun not work for me on angular 4 rc.5. Have any idea?

**Comment by kenese** [comment]
Hi there, I am also interested to know if we will be able to lazy load secondary routes in the future

**Comment by entrocode** [comment]
@vsavkin @jeffbcross We ran into this in your ng-conf session

**Comment by aniruddhadas9** [comment]
still facing the same issue.


http://stackoverflow.com/questions/40769797/angular2-invalid-route-configuration-of-route-outlate-name-a-componentless-ro?noredirect=1#comment73859966_40769797

**Comment by aherrick** [comment]
We are seeing this as well. We are simply trying to lazy load a module and have multiple named router outlets.

**Comment by jotatoledo** [comment]
Same issue here, I want to lazy load a module with a path to a named outlet. Cant figure a way to get it to work

**Comment by hajjem-ayoub** [comment]
Is there any information about this issue?? just so I know in my company if we should do a temporary solution until it's fixed or think about a more permanent solution. I just need an answer to a simple question:
"is this planned in the @angular/router team roadmap or not for now??"

**Comment by rsach** [comment]
i'm facing the similar issue , is there any work around for this?

**Comment by karsunke** [comment]
The workaround of @UnwrittenFun works for me in 4.2.5 (maybe in every version). Thanks alot.

**Comment by wosevision** [comment]
+1, trying to lazy-load a couple of heavier views in my sidebar. Seems a pretty reasonable use-case.. I'd be interested to hear if there's an actual technical difficulty in implementing this, because what I'm seeing here is a couple of easy workarounds and an error message that feels more like an oversight than intentional – the error that's thrown doesn't mention `loadChildren` and only checks whether there's a component attached to a named outlet.

**Comment by mchavaillaz** [comment]
I tried the workaround of @UnwrittenFun with the version 4.2.5 but it doesn't work.

Could somebody help me with the workaround? @karsunke ?

My test application is available in this repo --> https://github.com/mchavaillaz/angular-router-named-outlet-lazy-loaded

I have the erro message "Error: Cannot match any routes. URL Segment: 'module-one'"

**Comment by karsunke** [comment]
@mchavaillaz I created an issue in your repository, where we can discuss a solution: https://github.com/mchavaillaz/angular-router-named-outlet-lazy-loaded/issues/1

**Comment by mchavaillaz** [comment]
Ok I resolved my problem. A full example of module lazy loading displayed in a named router outlet is available here --> https://github.com/mchavaillaz/angular-router-named-outlet-lazy-loaded.

**Comment by George02** [comment]
Any update on this ?

**Comment by topikus** [comment]
Any update on this ?

**Comment by Jinkwon** [comment]
Any update on this ?

**Comment by rBull** [comment]
Any update on this ?

**Comment by gmanio** [comment]
Any update on this?

**Comment by infinite2009** [comment]
There seems to be no update. I get stuck in this.

**Comment by clarkmclean** [comment]
Any update on this?

**Comment by som30ind** [comment]
I am facing same issue with unnamed **router-outlet** with **Lazy Loaded Module**. It throws same error "Cannot read property 'routes' of undefined".

**Any fix for this?**

**Comment by erbsenkoenig** [comment]
Ich kehre zurück am 01.11.2017.

Ich werde Ihre Nachricht nach meiner Rückkehr beantworten.
In dringenden Fällen senden Sie bitte eine Kopie Ihrer E-Mail für
technische Angelegenheiten an entwicklung@arxes-tolina.de, ansonsten an
info@arxes-tolina.de. Ein anderer Mitarbeiter wird sich dann Ihrer E-Mail
annehmen.


Hinweis: Dies ist eine automatische Antwort auf Ihre Nachricht  "Re:
[angular/angular] Lazy loaded module in named outlet throws error (#12842)"
gesendet am 23.10.2017 07:17:38.

Diese ist die einzige Benachrichtigung, die Sie empfangen werden, während
diese Person abwesend ist.

**Comment by wosevision** [comment]
Though I don't want to speak on behalf of the Angular team, I feel like something needs to be said here: 

_This issue doesn't get closer to resolution in proportion to the number of comments that say "Any update on this"/"Any fix for this"/etc._ Thumbs-upping those comments doesn't magically go into some petition for the Angular team to finish this.

If you're stuck, **read the thread** – there's a workaround, which as the name suggests, will get you working. If you'd like to get updates on this, save some keystrokes by **clicking the 'Subscribe' button** – that allows you to _receive updates in your email_. You'll also receive updates every time someone comments "Any update on this?", so please, don't be that person.

Better yet, be the change you want to see in the world and [submit a pull request](https://github.com/angular/angular/blob/master/CONTRIBUTING.md).

**Comment by gmfun** [comment]
Solution proposed by @UnwrittenFun mostly works, but on refreshing the page with secondary route this throwing error
        Cannot read property 'routes' of undefined
I have tried remove lazy loading
        { path: "test", component: ModalComponent, outlet: 'modal'},

**Comment by mir4ef** [comment]
I think the error `cannot read property 'routes' of undefined` is for the primary outlet. I had the same issue on refresh, so I had to specify a route for the primary outlet as well:

```typescript
this.router.navigate([{ outlets: { primary: [ 'home' ], modal: [ 'login' ] } }], { skipLocationChange: true });
```

doing 
```typescript
this.router.navigate([ '/home', { outlets: { modal: [ 'login' ] } }], { skipLocationChange: true });
```
didn't throw the error, but it didn't populate the named router outlet

**Comment by blackholegalaxy** [comment]
As stated buy @gmfun, the solution by @UnwrittenFun works. But refreshing make the app crash. So I think this bug should really be addressed soon... Lazy load a module in a name router outlet is quite a common pattern in auxiliary routes...

**Comment by rohanrichards** [comment]
I'm using @UnwrittenFun workaround, but I'm having trouble getting this work for child routes that require params. Any tips?

I'm trying something like this:

	{ path: '', redirectTo: '/tabs/(profileoutlet:profile/5b1e0749f6f1f97634b68de2)', pathMatch: 'full' },
	{
		path: 'tabs', component: TabsComponent, children: [
			{
				path: 'profile', component: TabsProxyComponent, outlet: 'profileoutlet',
				children: [
					{ path: ":id", loadChildren: "./pages/profile/view-profile/view-profile.module#ViewProfileModule" }
				]
			},
		]
	}

My route is loading, however no params are making it to the component. I have tested the module outside of the tabs and can pass in params fine.

**Comment by SvenSchoene** [comment]
What worked for me is two things:  
- Always have an empty-path redirection on all levels.
- Use the full named-outlet-URL-syntax for redirection.  

After hours and hours of debugging this I'm too tired to explain this more clearly, so let me just leave you with an example route-configuration I use here:

```ts
export const routes: Routes = [
    {
        path: '',
        redirectTo: '/index',
        pathMatch: 'full',
    },
    {
      path: 'index',
      component: IndexIndexComponent,
      children: [
        {
            path: '',
            redirectTo: '/index/(main:newsfeed//header:newsfeed)',
            pathMatch: 'full',
        },
        {
            path: 'newsfeed',
            loadChildren: "c/newsfeed-index/newsfeed-index.module#NewsfeedIndexModule",
            outlet: 'main',
        },
        {
            path: 'newsfeed',
            loadChildren: "c/me-header-index/me-header-index.module#MeHeaderIndexModule",
            outlet: 'header',
        },
      ]
    },
  ]
```

To be honest, while it is super-unintuitive and documentation and examples around this are missing entirely, I guess it actually makes sense that the Angular router would need the specific outlet and route for that outlet specified.
...or I'm just too tired and I can't see how stupid it is that this is necessary.

In any case, I can keep developing my application now... (after some sleep, of course!)

**Comment by cguilhermef** [comment]
Great @SvenSchoene! It make sense to me. I've reached some situations where an empty-path was necessary to handle with secondary routes. I think you found the answer more clearly: secondary routes - or named outlets - need a path specified (in this case, an empty path) to the primary outlet .

**Comment by nasreddineskandrani** [comment]
I don't like people that talk in the air. If you have a working solution please take the time to reproduce the initial example of the issue and fix it then post it online.

@UnwrittenFun solution is not working [in this stackblitz online example](https://stackblitz.com/edit/angular-lazyload-header-and-content?file=src%2Fapp%2Fapp.module.ts)

My example is the same as the initial issue post.

if you have a solution please edit my online example and repost it -> in other case i suggest highly to stop polluting the thread with unrelated cases and open new issues (my 2 cents).

**Comment by thoussem** [comment]
Same here, any updates ?

**Comment by maxime1992** [comment]
Found a [PR that fixes it](https://github.com/angular/angular/pull/25483).

And I've written a [comment](https://github.com/angular/angular/pull/25483#issuecomment-495249672) to show how to apply that patch locally...

Other than that nop :man_shrugging:

**Comment by asvishnyakov** [comment]
Any updates on this?

**Comment by asvishnyakov** [comment]
@molcik @maxime1992 I created PR with my own fix for that issue, may you please review it?

**Comment by DenysVuika** [comment]
Unfortunately, the issue is still there. Had to struggle finding workarounds for Angular 7 apps and most lead to switching off the lazy loading.

**Comment by atscott** [comment]
Confirmed still an issue in v9: https://stackblitz.com/edit/angular-ivy-ydvhpx?file=src%2Fapp%2Fapp.module.ts

**Comment by wakonp** [comment]
A temporary workaround would be to use the preloadingStrategy. 

```
@NgModule({
  imports: [
    RouterModule.forRoot(routes, {
      preloadingStrategy: PreloadAllModules,
    }),
  ],
  exports: [RouterModule],
})

**Comment by destus90** [comment]
@wakonp 
A temporary workaround is not to use lazy loading at all :-)

**Comment by DenysVuika** [comment]
Given the issue created 13 Nov 2016, that's not a **temporary** workaround.

**Comment by nasreddineskandrani** [comment]
@DenysVuika here is a solution: https://medium.com/front-end-tricks/dynamic-header-bars-and-or-contents-for-your-ivy-angular-app-56e1e7e3fc22

**Comment by Zombitch** [comment]
O.M.G... I just waste so much time to search why my lazy loading routing outlet was not working ...
I followed @nasreddineskandrani to "fix" it.

**Comment by raj2033** [comment]
@DenysVuika  Tried, didn't work for me.

**Comment by raj2033** [comment]
@kenese its 3yrs down the line now. and the answer is sadly 'NO'.

**Comment by atscott** [comment]
Reverted in https://github.com/angular/angular/commit/8f24bc9443b3872fe095d9f7f77b308a361a13b4. [Fails internal tests](https://sponge.corp.google.com/target?id=14096760-7e52-4ba4-9d4e-7224b4c19e84&target=//cloud/console/integration/tests/shell/navigation:redirect_hermetic_chrome-linux&searchFor=&show=FAILED&sortBy=STATUS).

**Comment by destus90** [comment]
The happiness was very close.

**Comment by lincolnthree** [comment]
Omg a fix!!! Yessss! I'll be able to delete soooooo much portal code.

**Comment by cherryland** [comment]
> @atscott Reverted in 8f24bc9. [Fails internal tests](https://sponge.corp.google.com/target?id=14096760-7e52-4ba4-9d4e-7224b4c19e84&target=//cloud/console/integration/tests/shell/navigation:redirect_hermetic_chrome-linux&searchFor=&show=FAILED&sortBy=STATUS).

Why should this be a community problem? This is a Google problem. Point. We applied the patch manually to @angular/router and ... uhm well it just works. This the reason why Angular is the devil. This project is too dependent on Google's death star.

**Comment by delego-wlritchi** [comment]
> > @atscott Reverted in [8f24bc9](https://github.com/angular/angular/commit/8f24bc9443b3872fe095d9f7f77b308a361a13b4). [Fails internal tests](https://sponge.corp.google.com/target?id=14096760-7e52-4ba4-9d4e-7224b4c19e84&target=//cloud/console/integration/tests/shell/navigation:redirect_hermetic_chrome-linux&searchFor=&show=FAILED&sortBy=STATUS).
> 
> Why should this be a community problem? This is a Google problem. Point. We applied the patch manually to @angular/router and ... uhm well it just works. This the reason why Angular is the devil. This project is too dependent on Google's death star.

If a change to Angular causes *any* consumer's tests to fail, then it's a backwards-incompatible change and needs to be reverted, or at the very least only included in a new major release with suitable warnings around it. In this case, Google just happens to be a high-value consumer of the library. The fact that their tests caught an issue before it made it into a release candidate is good for other consumers like us. See discussion on this topic on the PR #38379.

There's a separate argument to be made that it might be helpful to describe the cause of the test failures publicly, so that non-Google contributors could help with the fix, but that discussion is already happening on the PR and it seems like Google does intend to circle back with that information when they've identified the root cause.

**Comment by acohenOT** [comment]
@atscott can you please explain why this fix is listed in the change log for 10.1.0 but this issue is still open and I don't think the PR was merged. Is this is a mistake?

https://github.com/angular/angular/blob/master/CHANGELOG.md

"router: support lazy loading for empty path named outlets"

**Comment by atscott** [comment]
> @atscott can you please explain why this fix is listed in the change log for 10.1.0 but this issue is still open and I don't think the PR was merged. Is this is a mistake?
> 
> https://github.com/angular/angular/blob/master/CHANGELOG.md
> 
> "router: support lazy loading for empty path named outlets"

Yes, that’s a mistake. The commit was reverted because we found an issue after it went in. I’ve since investigated the failure more and reworked the solution. I plan to merge the change next week and it’ll go into the 11.0.0-next.1 release. 

Would you like to send a PR to correct the changelog?

**Comment by acohenOT** [comment]
Yes, will do.
